x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -175,6 +177,45 @@ protected static List<Batch<LogicalPlan>> rules() {
         return asList(substitutions, operators(), skip, cleanup(), defaultTopN, label);
     }
 
+    static class InvertSpatialFunctions extends OptimizerRules.OptimizerRule<UnaryPlan> {
+
+        InvertSpatialFunctions() {
+            super(TransformDirection.UP);
+        }
+
+        @Override
+        protected LogicalPlan rule(UnaryPlan plan) {
+            if (plan instanceof Filter filter) {
+                var condition = filter.condition().transformDown(SpatialRelatesFunction.class, this::invertIfNecessary);
+                if (filter.condition().equals(condition) == false) {
+                    plan = new Filter(filter.source(), filter.child(), condition);
+                }
+            }
+            if (plan instanceof Eval eval) {
+                List<Alias> fields = eval.fields();
+                List<Alias> changed = fields.stream()
+                    .map(f -> (Alias) f.transformDown(SpatialRelatesFunction.class, this::invertIfNecessary))
+                    .toList();
+                if (changed.equals(fields) == false) {
+                    plan = new Eval(eval.source(), eval.child(), changed);
+                }
+            }
+            return plan;
+        }
+
+        private SpatialRelatesFunction invertIfNecessary(SpatialRelatesFunction srf) {
+            if (srf.isCommutative()) {
+                // Never bother to invert commutative functions
+                return srf;
+            }
+            if (srf.left().foldable() && srf.right().foldable() == false) {
+                // If we have a constant on the left, and a field on the right, invert the function
+                return srf.invert();
```

Comment:
This looks very close to `LiteralsOnTheRight`. Is there a chance we can just use that one? E.g. by having `SpatialRelatesFunction` inherit from `BinaryOperator` rather than from `BinaryScalarFunction` (that may be a big ask, though, not sure that would work); or by generalizing `LiteralsOnTheRight` to handle both cases?

Reply:
I did investigate BinaryOperator for the original SpatialIntersects function, but it turns out BinaryOperator has a lot of assumptions that are not applicable here, so I backed out of that approach. I'll take a peek at LiteralsOnTheRight, to see if it is amenable to this.